resolwe/flow/models/entity.py

```diff
@@ -30,6 +58,59 @@ class Meta(BaseCollection.Meta):
     #: entity type
     type = models.CharField(max_length=100, db_index=True, null=True, blank=True)
 
+    #: duplication date and time
+    duplicated = models.DateTimeField(blank=True, null=True)
+
+    def is_duplicate(self):
+        """Return True if entity is a duplicate."""
+        return bool(self.duplicated)
+
+    def _duplicate(self, contributor, inherit_collection_membership):
+        """Duplicate (make a copy)."""
+        duplicate = Entity.objects.get(id=self.id)
+        duplicate.pk = None
+        duplicate.slug = None
+        name_max_length = Entity._meta.get_field('name').max_length
+        duplicate.name = 'Copy of {}'.format(self.name)[:name_max_length]
+        duplicate.contributor = contributor
+        duplicate.duplicated = datetime.datetime.now()
+        duplicate.save(force_insert=True)  # pylint: disable=protected-access
+
+        # TODO
+        # Create duplication data migration history entry.
+
+        assign_contributor_permissions(duplicate)
+
+        # Fields to inherit from original entity.
+        duplicate.created = self.created
+        duplicate.save()
+
+        # Duplicate entity's data objects.
+        # pylint: disable=no-member
+        orig_and_duplicated_data = self.data.all().duplicate(contributor, return_mapping=True)
```

Comment:
`orig_and_duplicated_data` -> `copied_data`

Reply:
And considering the comment bellow you can greatly simplify it with omitting `return_mapping=True`.